src/lib/operators/table_scan_impl.cpp

```diff
@@ -0,0 +1,246 @@
+
+#include "table_scan.hpp"
+
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "resolve_type.hpp"
+#include "storage/dictionary_column.hpp"
+#include "storage/fitted_attribute_vector.hpp"
+#include "storage/reference_column.hpp"
+#include "storage/table.hpp"
+#include "storage/value_column.hpp"
+
+namespace opossum {
+
+template <typename T>
+std::shared_ptr<const Table> TableScan::TableScanImpl<T>::on_execute() {
+  // create output table
+  auto result_table = std::make_shared<Table>();
+
+  // check if the input already contains reference columns (e.g., already was a scan)
+  auto deref_table = _in_table;
+  auto is_reference = false;
+
+  auto ref_col = std::dynamic_pointer_cast<ReferenceColumn>(_in_table->get_chunk(ChunkID{0}).get_column(_column_id));
+  if (ref_col != nullptr) {
+    // if the input contains reference columns, dereference them
+    is_reference = true;
+    deref_table = ref_col->referenced_table();
+  }
+
+  // create PosList and fill with values from scan
+  _create_pos_list(is_reference);
+
+  // create chunk (we will have a single chunk for the results)
+  auto chunk = Chunk{};
+
+  // Fill the chunk with reference columns to mirror the original relation
+  for (auto col_id = ColumnID{0}; col_id < deref_table->col_count(); ++col_id) {
+    // add column definition
+    // TODO(team): is there a nicer way to combine add_column_definition and emplace_chunk ?
+    result_table->add_column(deref_table->column_name(col_id), deref_table->column_type(col_id));
```

Comment:
Maybe you could have used add_column_definition.
Is there a reason why you don't use this method?

Reply:
In one of the earlier sprints, we added an `is_instantiated` flag for each column definition, which is only set to `true` if we call 'add_column`, but not if we only `add_column_definition` - thus, we used `add_column` in this case, but weren't quite happy with that